src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private T[] arrayList;
+    int size = 0;
+    int emptyPlaces = 10;
+
+    public ArrayList() {
+        arrayList = (T[]) new Object[emptyPlaces];
+    }
+
+    public void enlarge() {
+        emptyPlaces = (int) (size * 0.5);
+        T[] temporaryEnlargedArrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            temporaryEnlargedArrayList[i] = arrayList[i];
+        }
+        arrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            arrayList[i] = temporaryEnlargedArrayList[i];
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (emptyPlaces == 0) {
+            enlarge();
+        }
+        arrayList[size] = value;
+        emptyPlaces--;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("given index does not exist");
+        } else if (index == size - 1) {
+            set(value, size -1);
+        } else {
+            T[] temporaryArrayList = (T[]) new Object[size - index];
+            for (int i = index; i < size; i++) {
+                temporaryArrayList[i - index] = arrayList[i];
+            }
+            set(value, index);
+            for (int i = size - 1; i > index; i--) {
```

Comment:
Removing and then adding elements back in is inefficient. Use `System.arraycopy()` to shift the elements once instead of removing and adding them one by one.